tensornetwork/contractors/stochastic_contractor.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2019 The TensorNetwork Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Stochastic Network Contraction."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import random
+import numpy as np
+from tensornetwork import tensornetwork
+
+
+def get_random_edge(network: tensornetwork.TensorNetwork) -> tensornetwork.Edge:
+  """Returns a random connected edge.
+  If it finds a trace edge it contracts it.
+  Args:
+    network: Connected TensorNetwork we are trying to contract.
+  Returns:
+    edge: A random connected edge from the network.
+  """
+  node = random.choice(tuple(network.nodes_set))
+  edge = random.choice(node.edges)
+  while edge.is_dangling():
+    edge = random.choice(node.edges)
+  if edge.node1 is edge.node2:
+    # Contract the trace edge and look for a new one
+    network.contract(edge)
+    return get_random_edge(network)
+  return edge
+
+
+def edge_cost(edge: tensornetwork.Edge) -> int:
+  """Calculates cost of contracting an edge.
+  If A and B are the tensors that share the given `edge`, cost is defined as:
+  cost = dims(A * B) - max(dims(A), dims(B)), where
+  * denotes contraction of all shared edges (`contract_between`) and
+  dims(X) is the total dimension of tensor X (product of sizes of all axes).
+  Args:
+    edge: Edge to calculate the cost.
+  Returns:
+    cost: Cost of the given edge.
+  """
+  # TODO: Verify whether this loss makes sense.
+  node1, node2 = edge.node1, edge.node2
+  # Calculate dimension of all shared edges
+  nodes = {node1, node2}
+  shared_dimension = 1
+  for edge in node1.edges:
+    if set(edge.get_nodes()) == nodes:
+      shared_dimension *= int(edge.node1.get_tensor().shape[edge.axis1])
+
+  dimension1 = int(np.prod(node1.get_tensor().shape))
+  dimension2 = int(np.prod(node2.get_tensor().shape))
+  prod_dimension = ((dimension1 // shared_dimension) *
+                    (dimension2 // shared_dimension))
+  cost = prod_dimension - max(dimension1, dimension2)
+  return cost
+
+
+def stochastic(network: tensornetwork.TensorNetwork,
+               max_rejections: int, threshold: int = -1
```

Comment:
Our cost is determined by the memory usage of a contraction. So with that, let's make the default threshold something larger like 128MB (~2**26)?

Reply:
The exponential *2 (previous comment) argument sounds good and I think it is in agreement with the paper. However I think that 2**26 might be huge for most cases. For example there are cases where cost < 0 (when the result of the contraction is smaller (in terms of memory) than the larger of the two tensors - these are the ones you would like to contract first I think). There is nothing wrong with using large threshold but then the order of contractions will be totally random (the cost calculation won't matter). I think it is hard to come up with a good universal default value because it depends a lot on the tensor sizes.

Reply:
We could set it as `None` as default to be the size of the largest tensor in the initial network. I think that is a good compromise.

Reply:
Calculating a default value according to tensor sizes is a good idea. I am not sure if the largest is the best choice (vs an average for example) because most random moves will still be accepted this way, but we can check that. The only issue is that we have to calculate all sizes prior starting so it would be more efficient to keep them in a map for future cost calculations. I started implementing this but currently it gets a bit messy because I have to update the map after every contraction. That's just a heads up that the code might change a lot to do this.

Reply:
No problem! Keep doing the good work and let me know when you are ready.